File: 310.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        adjList = [set() for _ in range(n)]
        for e in edges:
            adjList[e[0]].add(e[1])
            adjList[e[1]].add(e[0])
        leafNodes = [i for i, l in enumerate(adjList) if len(l) == 1]
        while n > 2:
            newLeaves = []
            for leaf in leafNodes:
                j = adjList[leaf].pop()
                adjList[j].remove(leaf)
                if len(adjList[j]) == 1:
                    newLeaves.append(j)
            n -= len(leafNodes)
            leafNodes = newLeaves
        return leafNodes

File: sol/310/310.cpp
#include <deque>

struct Solution {
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        if (n == 1){
            return vector<int>{0};
        }
        vector<vector<int>> adjList(n);
        vector<int> deg(n);
        for (auto &e : edges){
            adjList[e[0]].push_back(e[1]);
            adjList[e[1]].push_back(e[0]);
            ++deg[e[0]];
            ++deg[e[1]];
        }
        deque<int> leafNodes;
        for (int i = 0; i < n; ++i){
            if (adjList[i].size() == 1){
                leafNodes.push_back(i);
            }
        }
        while (n > 2){
            int s = leafNodes.size();
            n -= s;
            while (s-- > 0){
                int leaf = leafNodes.front();
                leafNodes.pop_front();
                deg[leaf] = 0;
                for (int nb : adjList[leaf]){
                    if (deg[nb] == 0){
                        continue;
                    }
                    --deg[nb];
                    if (deg[nb] == 1){
                        leafNodes.push_back(nb);
                    }
                }
            }
        }
        return vector<int>(leafNodes.begin(), leafNodes.end());
    }
};
